Mage.Sets/src/mage/cards/k/KaghaShadowArchdruid.java

```diff
@@ -0,0 +1,168 @@
+    }
+}
+
+class KaghaShadowArchdruidEffect extends AsThoughEffectImpl {
+
+    private final FilterCard filter;
+
+    KaghaShadowArchdruidEffect(FilterCard filter, String text) {
+        super(AsThoughEffectType.PLAY_FROM_NOT_OWN_HAND_ZONE, Duration.WhileOnBattlefield, Outcome.Benefit, true);
+        this.filter = filter;
+        this.staticText = text;
+    }
+
+    private KaghaShadowArchdruidEffect(final KaghaShadowArchdruidEffect effect) {
+        super(effect);
+        this.filter = effect.filter;
+    }
+
+    @Override
+    public KaghaShadowArchdruidEffect copy() {
+        return new KaghaShadowArchdruidEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        return true;
+    }
+
+    @Override
+    public boolean applies(UUID objectId, Ability source, UUID affectedControllerId, Game game) {
+        if (source.isControlledBy(affectedControllerId)
+                && Zone.GRAVEYARD.equals(game.getState().getZone(objectId))
+                && game.isActivePlayer(affectedControllerId)) {
+            Card card = game.getCard(objectId);
+            Permanent sourceObject = source.getSourcePermanentIfItStillExists(game);
+            if (card != null && sourceObject != null 
+                    && card.isOwnedBy(affectedControllerId)
+                    && filter.match(card, affectedControllerId, source, game)) {
+                KaghaShadowArchdruidWatcher watcher = game.getState().getWatcher(KaghaShadowArchdruidWatcher.class);
+                return watcher != null && watcher.abilityNotUsed(new MageObjectReference(sourceObject, game));
+            }
+        }
+        return false;
+    }
+}
+
+class KaghaShadowArchdruidWatcher extends Watcher {
+
+    private final Set<MageObjectReference> usedFrom = new HashSet<>();
+    private final Set<MageObjectReference> morSet = new HashSet<>();
+
+    KaghaShadowArchdruidWatcher() {
+        super(WatcherScope.GAME);
+    }
+
+    @Override
+    public void watch(GameEvent event, Game game) {
+        if ((GameEvent.EventType.SPELL_CAST.equals(event.getType()) || GameEvent.EventType.LAND_PLAYED.equals(event.getType()))
+                && event.hasApprovingIdentifier(MageIdentifier.KaghaShadowArchdruidWatcher)) {
+            usedFrom.add(event.getAdditionalReference().getApprovingMageObjectReference());
+        }
+        if (event.getType() == GameEvent.EventType.ZONE_CHANGE) {
+            ZoneChangeEvent zEvent = (ZoneChangeEvent) event;
+            if (zEvent.getFromZone() == Zone.LIBRARY && zEvent.getToZone() == Zone.GRAVEYARD) {
+                morSet.add(new MageObjectReference(zEvent.getTargetId(), game));
+            }
+        }
+    }
+
+    @Override
+    public void reset() {
+        super.reset();
+        usedFrom.clear();
+        morSet.clear();
+    }
+
+    boolean abilityNotUsed(MageObjectReference mor) {
+        return !usedFrom.contains(mor);
+    }
+
+    static boolean checkCard(Card card, Game game) {
```

Comment:
(this can become non-static when not being used in a filter)

Reply:
Done

Reply:
Sorry, I missed that comment yesterday.
I am not sure if I understand you correctly. I can move the `card.isPermanent()` check to the `applies()` method but `morSet` is private therefore I need a function to access this (like `abilityNotUsed()`).

Reply:
For package-private class (without an explicit access keyword public/protected/private) like the new Watcher, there is no harm to have package-private of its attribute.

Responded there also https://github.com/magefree/mage/pull/10919#discussion_r1301073199

Reply:
Currently you have
```
    boolean checkCard(Card card, Game game) {
        return card.isPermanent(game) && game
                .getState()
                .getWatcher(KaghaShadowArchdruidWatcher.class)
                .morSet
                .stream()
                .anyMatch(mor -> mor.refersTo(card, game));
    }
```
You're accessing morSet not from the instance itself but from the instance in the GameState.
This would be the way for static methods used multiple places but overkill here. Instead use

```
    boolean checkCard(Card card, Game game) {
        return morSet
                .stream()
                .anyMatch(mor -> mor.refersTo(card, game));
    }
```
Make sense?

Reply:
That makes sense 👍 
I completely overlooked that I missed to remove this leftover from the static method. Change is incomming